### PR TITLE
Reduce the number of Rust call in JS side - Closes #35

### LINKED ISSUE #35: Reduce a number of Rust function calls from JavaScript code
### Description

- when calling Rust bindings from JavaScript code we get some execution performance penalty due to an additional code which needs to be executed because of a bridge between two different programming languages
- A good example are functions `StateReadWriter.set()` and `StateReadWriter.get()` in _state_db.js_ file. In these two functions there are multiple Rust binding calls. Because these two functions are heavily used, overall code execution performance is slow because of a lot of bridge calls between JavaScript and Rust code.
- develop an idea how such JavaScript functions could have less Rust function calls

### Acceptance Criteria

- JavaScript code has less Rust function calls

### Additional Information

- one idea is that JavaScript part of _lisk-db_ will have it's own temporary database/cache to store and read data inside it and only call Rust code in case that expected data won't be present in its database/cache

### PR DESCRIPTION
**Pull request by hrmhatef**
### What was the problem?

This PR resolves #35 

### How was it solved?
 - Reduce the number of rust call to one for each function in `JS` side

### How was it tested?
 - Compare the speed of the execution between old implementation and new one, post the result on `slack` channel.
 - All tests passed.

### CODE REVIEW

File: src/state_db.rs [hunk 1]
```diff
@@ -454,9 +451,293 @@ impl StateDB {
...
+                        if value.is_some() {
+                            let temp_value = value.as_ref().unwrap().to_vec();
+                            let pair = SharedKVPair::new(&key, &temp_value);
+                            writer.cache_existing(&pair);
+                            let result = writer.update(&KVPair::new(&key, &new_value));
+                            Self::parse_update_result(&mut ctx, result, 2)?
+                        } else {
+                            // if there is no key then make a new pair and insert into cache
+                            writer.cache_new(&SharedKVPair::new(&key, &new_value));
+                            vec![ctx.null().upcast(), JsNumber::new(&mut ctx, 3).upcast()]
```
Comment (matjazv) [inline src/state_db.rs:501]: What numbers 1, 2 and 3 represents? Should they be group in enum?

File: src/state_db.rs [hunk 2]
```diff
@@ -454,9 +451,293 @@ impl StateDB {
...
+        writer: ArcMutex<state_writer::StateWriter>,
+        key: Vec<u8>,
+        new_value: Vec<u8>,
+    ) -> Result<(), mpsc::SendError<options::DbMessage>> {
+        let state_db_key = DBKind::State.key(key.clone());
+        self.common.send(move |conn, channel| {
+            let value = conn.get(&state_db_key);
+            channel.send(move |mut ctx| {
+                let args = {
+                    let mut writer = writer.lock().unwrap();
```
Comment (matjazv) [inline src/state_db.rs:484]: You may release writer in similar way which is done in `delete_key` function.

File: src/state_db.rs [hunk 3]
```diff
@@ -454,9 +451,293 @@ impl StateDB {
...
+        cb: Root<JsFunction>,
+        writer: ArcMutex<state_writer::StateWriter>,
+        key: Vec<u8>,
+    ) -> Result<(), mpsc::SendError<options::DbMessage>> {
+        let state_db_key = DBKind::State.key(key.clone());
+        self.common.send(move |conn, channel| {
+            let value = conn.get(&state_db_key);
+            channel.send(move |mut ctx| {
+                let args = {
+                    let mut writer = writer.lock().unwrap();
```
Comment (matjazv) [inline src/state_db.rs:567]: You may release writer in similar way which is done in `delete_key` function.

File: src/state_db.rs [hunk 4]
```diff
@@ -454,9 +451,293 @@ impl StateDB {
...
+        options: options::IterationOption,
+    ) -> Result<(), mpsc::SendError<options::DbMessage>> {
+        self.common.send(move |conn, channel| {
+            let values = conn
+                .iterator(utils::get_iteration_mode(&options, &mut vec![], true))
+                .map(|(k, v)| KVPair::new(&k, &v))
+                .collect::<Vec<KVPair>>();
+            channel.send(move |mut ctx| {
+                let result = {
+                    let mut writer = writer.lock().unwrap();
```
Comment (matjazv) [inline src/state_db.rs:639]: You may release writer in similar way which is done in `delete_key` function.

**Review by matjazv — CHANGES_REQUESTED** [review]

File: src/state_db.rs (hunk 1 above)
Comment (hrmhatef) [reply src/state_db.rs:501]: I used these values for debugging and better understanding of what happen. So one I can remove them or change them into enum for the future. if you think the code is clear with comments so it is better to remove them.

File: src/state_db.rs (hunk 3 above)
Comment (hrmhatef) [reply src/state_db.rs:567]: Same as above comment

File: src/state_db.rs (hunk 4 above)
Comment (hrmhatef) [reply src/state_db.rs:639]: same as `args`

File: src/state_db.rs (hunk 1 above)
Comment (hrmhatef) [reply src/state_db.rs:501]: Numbers removed.

**Review by matjazv — APPROVED** [review]

**Review by shuse2 — APPROVED** [review]

**Pull request merged**